shared/common-adapters/oriented-image.desktop.js

```diff
@@ -0,0 +1,71 @@
+// @flow
+import * as React from 'React'
+import EXIF from 'exif-js'
+import {noop, inRange} from 'lodash-es'
+import {Image} from '../common-adapters'
+import {type Props} from './oriented-image.types'
+import {collapseStyles, type StylesDesktop} from '../styles'
+
+type State = {
+  styleTransform: StylesDesktop | null,
+}
+
+const makeStyleTransform = (orientation: number): StylesDesktop =>
+  exifOrientaionMap[orientation] || {transform: 'rotate(0deg)'}
+
+// Orientations:
+// 1: rotate 0 deg left
+// 2: flip horizontally
+// 3: rotate 180 deg left
+// 4: flip vertivally
+// 5: flip vertivally and rotate 270 degrees left
+// 6: rotate 270 deg left
+// 7: flip vertically and rotate 90 degrees left
+// 8: rotate 90 deg left
+const exifOrientaionMap = {
+  '1': {transform: 'rotate(0deg)'},
+  '2': {transform: 'scale(-1, 1)'},
+  '3': {transform: 'rotate(180deg)'},
+  '4': {transform: 'scale(1, -1)'},
+  '5': {transform: 'scale(1, -1) rotate(270deg)'},
+  '6': {transform: 'rotate(270deg)'},
+  '7': {transform: 'scale(1, -1) rotate(90deg)'},
+  '8': {transform: 'rotate(90deg)'},
+}
+
+class OrientedImage extends React.Component<Props, State> {
+  static defaultProps = {
+    onLoad: noop,
+  }
+  state = {
+    styleTransform: null,
+  }
+  _handleOnLoad = () => {
+    if (this.state.styleTransform !== null) {
+      this.props.onLoad()
+    }
+  }
+  componentDidMount() {
+    const component = this
+    EXIF.getData({src: this.props.src}, function() {
+      const img = this
+      const orientation: number = EXIF.getTag(img, 'Orientation')
+      if (orientation && inRange(orientation, 1, 9)) {
+        const transform: StylesDesktop = makeStyleTransform(orientation)
+        component.setState({styleTransform: transform})
```

Comment:
Not entirely happy with calling `setState()` inside of `componentDidMount` -> I've always understood this to be an anti-patter and possibly leading to two renders.

Seeing as we only want to perform this once, the other thought I had was to move the EXIF library call to a parent container and hold off the loading state until it's completed.

Looking for feedback on this.

Reply:
lets chat about this

Reply:
^ chatted